core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java

```diff
@@ -80,38 +80,43 @@ public void append(IntArrayList positions, Block block)
             return;
         }
         ensurePositionCapacity(positionCount + positions.size());
-        int[] positionArray = positions.elements();
-        int newByteCount = 0;
-        int[] lengths = new int[positions.size()];
-
-        if (block.mayHaveNull()) {
-            for (int i = 0; i < positions.size(); i++) {
-                int position = positionArray[i];
-                if (block.isNull(position)) {
-                    offsets[positionCount + i + 1] = offsets[positionCount + i];
-                    valueIsNull[positionCount + i] = true;
-                    hasNullValue = true;
+        if (block instanceof VariableWidthBlock) {
+            VariableWidthBlock variableWidthBlock = (VariableWidthBlock) block;
+            int newByteCount = 0;
+            int[] lengths = new int[positions.size()];
+            int[] sourceOffsets = new int[positions.size()];
+            int[] positionArray = positions.elements();
+
+            if (block.mayHaveNull()) {
+                for (int i = 0; i < positions.size(); i++) {
+                    int position = positionArray[i];
+                    int length = variableWidthBlock.getSliceLength(position);
+                    lengths[i] = length;
+                    sourceOffsets[i] = variableWidthBlock.getRawSliceOffset(position);
+                    newByteCount += length;
+                    boolean isNull = block.isNull(position);
```

Comment:
just a thought:
We can add Slice#getNulls(boolean[] to, int offset, int length) or sth like that to copy the entire portion of isNull array and get rid of bound checks this way. 
But this is clearly out of scope here. Just wanted to write it down somewhere

Reply:
we would have to then pass `positions` to the `block.getNulls` (positions are not consecutive).
I did experiments with this idea but I don't see any improvement